src/citrine/resources/dataset.py

```diff
@@ -345,7 +348,31 @@ def register(self, model: Dataset) -> Dataset:
         path = self._get_path()
         dumped_dataset = model.dump()
         dumped_dataset["deleted"] = None
-        data = self.session.post_resource(path, scrub_none(dumped_dataset))
+
+        # Only use the idempotent put approach if a) a unique name is provided, and b)
+        # the session is configured to use it (default to False for backwards compatibility).
+        if model.unique_name is not None and self.session.use_idempotent_dataset_put:
+            # Leverage the create-or-update endpoint if we've got a unique name
+            data = self.session.put_resource(path, scrub_none(dumped_dataset))
```

Comment:
Is this the correct step if a user has accidentally stumbled upon someone else's `unique_name`?  This seems like a particular risk if there's no way for a user to ask the platform what `unique_name`s have been used.

Reply:
The POST endpoint will fail if the unique name is already in use. The PUT will update. Currently it's flagged off, but this is the intended behavior.

Reply:
Dj is right. As of right now there are no datasets with a unique name, because the migration defaults to null. If a user tries to create (Post) two new distinct datasets with the same unique name, they'll get an error on the second one. That's by design.

For the create-or-update (Put) case, we specifically choose to update the given dataset on the second call (as identified by the unique name). This maps to the intended contract of .register(). We're just not using this by default because of backwards compatibility concerns. I imagine we'll use that by default at some stage (assuming the unique name is provided).

Reply:
So what happens once this is deployed, and the second user tries to create a dataset with `unique_name = "my_dataset"`?  What is actually at risk of being clobbered? As best as I can think, this might mess up `name`, `summary` and `description`, but should be able to impact dataset ownership or access control, right?

Reply:
An update won't change any grants, that only happens on creation (and is the same for POST vs PUT).

Yeah, the second write will overwrite the first one with the supplied properties. The actual query updates: name, summary, description, updated_by and updated_time. The UUID, unique_name, and all grants will stay the same.

Reply:
And just for the curious: when we do an update on a dataset we don't have WRITE access on - we actually perform the UPSERT operation, get back the ID it's using, and check for Write access after the fact. If that fails, then the whole transaction gets rolled back and no updates are made. Kind of cool actually.

Reply:
@kroenlein is it your contention (it seems like it is) that people should be able to create a dataset knowing only the unique name, but that they should only be able to PUT / UPDATE a dataset using an ID? That seems like a user-hostile constraint to me. If a user wants to be certain their chosen `unique_name` is available, they can do a GET on the unique name before attempting a PUT, or they can use the POST method. PUT/UPDATE should operate as exactly that - createOrUpdate.

Reply:
I'm just trying to figure out if we've got a nasty tripping hazard.  Stumbling across an existing unique_name seems reasonably likely, so if the outcome would be a rename and objects in the wrong data set (hopefully discovered quickly), that's not so bad.  Ideally, a user trying to create a new dataset with an existing unique_name would bounce, but it's not obvious to me that we could differentiate between that case and the update case.  

Maybe if you want to update a dataset, you need to match both ID & unique_name?  It is a little user hostile, but a GET -> update -> PUT doesn't seem terrible if you want to edit a record on the platform.  I'm not gonna block on this.

Reply:
If we don't want to klobber the contents of an existing dataset, we can call a POST with that unique name, catch the error, and handle it appropriately. Perhaps the distinction between create and update should be called out in the client, as .register() conflates the two.

For now I think this is fine (there's no documented client-visible change yet), but maybe we should think about parameters to DatasetCollection.register, or other ways to prevent mistakes.


The main benefit for us is that IF we're confident that our unique name is accurate, then we have one call that creates or updates the dataset based on that single criteria. This is particularly important for parallel e2e tests because this call is atomic and guaranteed to give consistent results. That's probably not so important for a typical user though.